text/0000-glimmer-components.md

```diff
@@ -0,0 +1,1020 @@
+hooks, with as little an API surface as possible to learn and remember.
+
+### Element Modifiers
+
+Glimmer components have outer HTML semantics, which means that whatever is put
+in the template is what we see in the final render. This also means that they
+can have _multiple_ top-level elements:
+
+```hbs
+<div ...attributes>
+  First root element
+</div>
+<div>
+  Second root element
+</div>
+```
+
+It is also possible that there are _no_ root elements:
+
+```hbs
+This is a text only component: {{@foo}}
+```
+
+This makes the `element` property of classic components awkward, since there is
+no one single referenceable element. A previous proposal for solving this
+problem was a `bounds` property which contained `firstNode` and `lastNode`, but
+this could be very confusing to users since these nodes would likely have to be
+text nodes (how Glimmer tracks component boundaries internally), and would force
+users to write unintuitive looping logic to traverse nodes:
+
+```js
+let { firstNode, lastNode } = bounds;
+let currentNode = firstNode;
+
+while (currentNode !== lastNode) {
+  // do something with current node
+  currentNode = currentNode.nextSibling;
+}
+```
+
+This exposes users to low level APIs which are fairly complicated for most
+simple use cases, and can be error prone. We could also attempt to add an
+`element` property for cases where templates only have one root level element,
+but this represents a refactoring hazard if users attempt to rewrite code, and
+means we have to deal with a few complexities (what if a component has top level
+text nodes?)
+
+Instead, we can use new element modifiers which allow users to run code
+declaratively on the element:
+
+```hbs
+<div {{did-render this.setupElement}} ...attributes>
+  First root element
+</div>
+<div {{did-render this.setupElement}}>
+  Second root element
+</div>
+```
+
+This allows users to target which elements they care about directly. It also
+allows for code to be run on elements which are contained in conditionals,
+meaning users do not need to add messy logic to track component state
+themselves:
+
+```hbs
+{{#if shouldRender}}
+  <div
+    {{did-render this.setupElement}}
+    {{will-destroy this.teardownElement}}
+  >
+
+  </div>
+{{/if}}
+```
+
+Finally, it means that users can bypass component code altogether by using
+helpers:
+
+```hbs
+<div {{did-render setScrollPosition @scrollPosition}}></div>
+```
+
+This RFC proposes adding two element modifiers: `{{did-render}}` and
+`{{will-destroy}}`.
+
+#### `{{did-render}}`
+
+This modifier runs immediately after the element is inserted, and again whenever
+any of the arguments passed to it update, including the function passed as the
+first argument. It has the following timing semantics:
+
+* **Always**
+  * called _after_ the enclosing component's `didUpdateArgs` hook
+  * called _after_ any child element's `{{did-render}}` modifiers
+  * called in definition order in the template
```

Comment:
I can imagine some common footgun factor here, regarding refactoring templates (traditionally affecting only the _structure_ of the component and never the _behavior_) affecting the order of modifiers firing. 

If I were leading a team, I'd want people to author any "modifier-based hooks" as if there's no guaranteed order

Reply:
Added a section to learning that calls out that we should not focus on ordering, and that best practices for modifiers are to be commutative (order-independent)